third_party/nvfuser/csrc/scheduler/normalization.cpp

```diff
@@ -1221,4 +1515,257 @@ void schedulePersistentKernel(Fusion* fusion, const ReductionParams& rparams) {
+  // the new tensor view contains the reduction and original doesn't.
+
+  bool unroll = rparams.isUnrolled();
+
+  // Cache inputs even if not unrolled, as otherwise we may not create a
+  // persistent buffer if that persistent buffer would be the input.
+  auto cached_inputs = scheduler_utils::cacheInputs(fusion, true);
+
+  // Cache and fork outputs
+  auto cached_outputs = scheduler_utils::cacheAndForkOutputs(fusion, unroll);
+
+  // Make sure we don't have global memory set on intermediate tensors from
+  // fusion segmentation
+  scheduler_utils::clearMemorySpace(fusion);
+
+  auto persistent_info = scheduler_utils::persistentBuffers(fusion);
+
+  auto reduction_tvs = scheduler_utils::getReductionTvs(fusion);
+  /////////////////////////////////////////////////////////////////////////////////////////////////
+  std::vector<TensorView*> inner_reduction_tvs, outer_reduction_tvs;
+  for (auto tv : reduction_tvs) {
+    if (scheduler_utils::isFastestDimReduction(tv)) {
+      inner_reduction_tvs.emplace_back(tv);
+    } else {
+      outer_reduction_tvs.emplace_back(tv);
+    }
+  }
+  // inner reduction
+  auto inner_reduction_tv = inner_reduction_tvs[0];
+  if (ir_utils::getViewOps(fusion).size() > 0) {
+    ComputeAtMap ca_map(fusion);
+    // Propagate view transforms through the graph, expecially the reference.
+    scheduler_utils::propagateViewTransforms(fusion, ca_map);
+
+    // Reorder reference_tv after propagating the view operation. This will
+    // reorder for better merging.
+    inner_reduction_tv->reorder(
+        scheduler_utils::domainReorderAsRfactorMap(inner_reduction_tv));
+  }
+
+  auto dim_analysis = scheduler_utils::canonicalDimReduction(
+      fusion, inner_reduction_tv, rparams.fastest_dim && rparams.schedule_3D);
+  bool has_iter_axis = dim_analysis.first;
+  bool has_red_axis = dim_analysis.second;
+
+  TORCH_INTERNAL_ASSERT(
+      has_red_axis,
+      "Could not find reduction axis in tensor used for reduction scheduler.");
+
+  if (!has_iter_axis) {
+    TORCH_INTERNAL_ASSERT(
+        rparams.fastest_dim,
+        "If all dims are reduction, should be sending it to fastest dim scheduler.");
+  }
+
+  TensorView* inner_reference_tv =
+      reduction_scheduler_utils::scheduleReductionTV(
+          rparams, inner_reduction_tv, has_iter_axis);
+
+  // outer reduction
+  std::vector<TensorView*> cached_gmem;
+  std::vector<TensorView*> cached_gmem_reload;
+  std::vector<TensorView*> outer_reference_tvs;
+  std::unordered_set<TensorView*> boundaryNodesSet;
+  auto mergeReductionOrIterDomains = [](TensorView* tv, bool mergeReduction) {
+    int prev_i = -1;
+    for (int i = static_cast<int>(tv->nDims()) - 1; i >= 0; i--) {
+      if (mergeReduction == tv->axis(i)->isReduction()) {
+        if (prev_i == -1) {
+          prev_i = i;
+        } else {
+          tv->merge(i, prev_i);
+          prev_i = i;
+        }
+      }
+    }
+  };
+  for (auto& outer_reduction_tv : outer_reduction_tvs) {
```

Comment:
Do we need to schedule each of the outer reduction tvs? Can we just schedule one and propagate it to the rest?

Reply:
if we just schedule one, then we can't easily get the intermediate results/info, e.g. 
```
 std::vector<TensorView*> cached_gmem;
  std::vector<TensorView*> cached_gmem_reload;
  std::unordered_set<TensorView*> boundaryNodesSet;
```
added comments:
```
  // schedule outer reduction, schedule all the outer reduction tvs since we
  // need to store the intermediate results.
```